# **Automatidata Project**
**Examining and understanding the data**

## **Introduction**

Welcome to the Automatidata Project!

The automatidata project is one out of three project that are available at the end of each course of the Google Advance Data Analytics Certificate. This is one of the first lab project activity out of the seven course that are available. For this lab activity, I will be examining the data and preparing the data for analysis during the early stage of the project. There will be more labs as I advance through the course, by the end of the last labs there will be a complete pictures for the analysis project and tasks. The course introduce the PACE workflow and explained how it provides a foundation for data analysis projects. PACE stand for plan, analyze, construct, and execute. 

## **Scenario and Tasks**

Automatidata works with its clients to tranform their unused and stored data into usefull solutions, such as performance dashboards, customer-facing tools, strategic business insights, and more. They specialize in identifying a client's business needs and utilizing their data to meet those business needs. 

Automatidata is consulting for the New York City Taxi and Limousine Commission (TLC). New York City TLC is an agency responsible for licensing and regulating New York City's taxi cabs and for-hire vehicles. The agency has partnered with Automatidata to develop a regression model that helps estimate taxi fares before the ride, based on data that TLC has gathered.

The TLC data comes from over 200,000 taxi and limousine licensees, making approximately one million combined trips per day.

Automatidata is in the earliest stages of the TLC project. The following tasks are needed before the team can begin the data analysis process:

- Build a dataframe for the TLC dataset
- Examine data type of each column
- Gather descriptive statistics

## **Inspect and analyze data**

In this activity, I will be examining the data provied and prepare it for analysis. This activity will help ensure the information is:

1. Ready to answer questions and yield insights
2. Ready for visualizations
3. Ready for future hypothesis testing and statistical methods
<br/>

**The purpose** of this activity is to investigate and understand the data provided.

**The goal** is to use a dataframe constructed within Python, perform a cursory inspection of the provided dataset, and inform team members of our findings.
<br/>

**Part 1:** Understand the situation
* Prepare to understand and organize the provided taxi dataset and information.

**Part 2:** Understand the data
* Create a pandas dataframe for data learning, future exploratory data analysis (EDA), and statistical activities.
* Compile summary information about the data to inform next steps.

**Part 3:** Understand the variables
* Use insights from the examination of the data summary to guide deeper investigation into specific variables.

# **Identify data types and relevant variables**

## **PACE stages**

Throughout these project notebooks, there will be references to the problem-solving framework PACE.

## **PACE: Plan**

### **Understand the situation**

The data are provide for this project, best way to understand the data is by exploring the dataset and consider reviewing the Data Dictionary. One can prepare to understand the information by reading the taxi cab data fields and understanding the impact of each one.

Reviewing the fact sheet could also provide usefull background information. However, the primary goal is to get the data into Python, inspect it, and provide your team member/manager with initial observations. The nest step would be to learn more about the data and check for any anomalies.

## **PACE: Analyze**

### **Build dataframe**

Create a pandas dataframe for data learning, and future exploratory data analysis (EDA) and statistical activities.

**Importting data and packages:**
- import pandas for building dataframes
- import numpy
- read data from provide .csv

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv('2017_Yellow_Taxi_Trip_Data.csv')

### **Inspect the data**

View and inspect summary information about the dataframe using:

1. head()
2. info()
3. describe()

In [2]:
df.head()

Unnamed: 0  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
0    24870114         2   03/25/2017 8:55:43 AM   03/25/2017 9:09:47 AM   
1    35634249         1   04/11/2017 2:53:28 PM   04/11/2017 3:19:58 PM   
2   106203690         1   12/15/2017 7:26:56 AM   12/15/2017 7:34:08 AM   
3    38942136         2   05/07/2017 1:17:59 PM   05/07/2017 1:48:14 PM   
4    30841670         2  04/15/2017 11:32:20 PM  04/15/2017 11:49:03 PM   

   passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
0                6           3.34           1                  N   
1                1           1.80           1                  N   
2                1           1.00           1                  N   
3                1           3.70           1                  N   
4                1           4.37           1                  N   

   PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
0           100           231             1         13.0    0.0      0.5   
1           186            43             1         16.0    0.0      0.5   
2           262           236             1          6.5    0.0      0.5   
3           188            97             1         20.5    0.0      0.5   
4             4           112             2         16.5    0.5      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  
0        2.76           0.0                    0.3         16.56  
1        4.00           0.0                    0.3         20.80  
2        1.45           0.0                    0.3          8.75  
3        6.39           0.0                    0.3         27.69  
4        0.00           0.0                    0.3         17.80

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22699 entries, 0 to 22698
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             22699 non-null  int64  
 1   VendorID               22699 non-null  int64  
 2   tpep_pickup_datetime   22699 non-null  object 
 3   tpep_dropoff_datetime  22699 non-null  object 
 4   passenger_count        22699 non-null  int64  
 5   trip_distance          22699 non-null  float64
 6   RatecodeID             22699 non-null  int64  
 7   store_and_fwd_flag     22699 non-null  object 
 8   PULocationID           22699 non-null  int64  
 9   DOLocationID           22699 non-null  int64  
 10  payment_type           22699 non-null  int64  
 11  fare_amount            22699 non-null  float64
 12  extra                  22699 non-null  float64
 13  mta_tax                22699 non-null  float64
 14  tip_amount             22699 non-null  float64
 15  to

In [4]:
df.describe()

Unnamed: 0      VendorID  passenger_count  trip_distance  \
count  2.269900e+04  22699.000000     22699.000000   22699.000000   
mean   5.675849e+07      1.556236         1.642319       2.913313   
std    3.274493e+07      0.496838         1.285231       3.653171   
min    1.212700e+04      1.000000         0.000000       0.000000   
25%    2.852056e+07      1.000000         1.000000       0.990000   
50%    5.673150e+07      2.000000         1.000000       1.610000   
75%    8.537452e+07      2.000000         2.000000       3.060000   
max    1.134863e+08      2.000000         6.000000      33.960000   

         RatecodeID  PULocationID  DOLocationID  payment_type   fare_amount  \
count  22699.000000  22699.000000  22699.000000  22699.000000  22699.000000   
mean       1.043394    162.412353    161.527997      1.336887     13.026629   
std        0.708391     66.633373     70.139691      0.496211     13.243791   
min        1.000000      1.000000      1.000000      1.000000   -120.000000   
25%        1.000000    114.000000    112.000000      1.000000      6.500000   
50%        1.000000    162.000000    162.000000      1.000000      9.500000   
75%        1.000000    233.000000    233.000000      2.000000     14.500000   
max       99.000000    265.000000    265.000000      4.000000    999.990000   

              extra       mta_tax    tip_amount  tolls_amount  \
count  22699.000000  22699.000000  22699.000000  22699.000000   
mean       0.333275      0.497445      1.835781      0.312542   
std        0.463097      0.039465      2.800626      1.399212   
min       -1.000000     -0.500000      0.000000      0.000000   
25%        0.000000      0.500000      0.000000      0.000000   
50%        0.000000      0.500000      1.350000      0.000000   
75%        0.500000      0.500000      2.450000      0.000000   
max        4.500000      0.500000    200.000000     19.100000   

       improvement_surcharge  total_amount  
count           22699.000000  22699.000000  
mean                0.299551     16.310502  
std                 0.015673     16.097295  
min                -0.300000   -120.300000  
25%                 0.300000      8.750000  
50%                 0.300000     11.800000  
75%                 0.300000     17.800000  
max                 0.300000   1200.290000

**Quick Observation**

Looking at the info of the data, with a quick overview we see that there are no null values. All the variables seem to be numeric except for the two datetime variables which are shown as object and the date and time are combine together, we could seperate to create a new columns.

From the describe method summary, the fare amount seem to be questionable as the min is -120 and the max is 999.99 which need to be further investigate. Also, the maximum fare amount $1200 is a much large value than the 25-75 percent range of values.

Regarding trip distance, most rides are between 1-3 miles, but the maximum is over 33 miles.

### **Investigate the variables**

From the observation above, I want to look further into the trip duration and the total amounts of the taxi fare. Sort and interpret the data table for two variables:`trip_distance` and `total_amount`.

In [5]:
# sort the data by trip distance from maximum to minimum value

df_sort = df.sort_values(by = ['trip_distance'], ascending = False)
df_sort.head(10)

Unnamed: 0  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
9280     51810714         2  06/18/2017 11:33:25 PM  06/19/2017 12:12:38 AM   
13861    40523668         2   05/19/2017 8:20:21 AM   05/19/2017 9:20:30 AM   
6064     49894023         2  06/13/2017 12:30:22 PM   06/13/2017 1:37:51 PM   
10291    76319330         2  09/11/2017 11:41:04 AM  09/11/2017 12:18:58 PM   
29       94052446         2   11/06/2017 8:30:50 PM  11/07/2017 12:00:00 AM   
18130    90375786         1   10/26/2017 2:45:01 PM   10/26/2017 4:12:49 PM   
5792     68023798         2   08/11/2017 2:14:01 PM   08/11/2017 3:17:31 PM   
15350    77309977         2   09/14/2017 1:44:44 PM   09/14/2017 2:34:29 PM   
10302    43431843         1   05/15/2017 8:11:34 AM   05/15/2017 9:03:16 AM   
2592     51094874         2   06/16/2017 6:51:20 PM   06/16/2017 7:41:42 PM   

       passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
9280                 2          33.96           5                  N   
13861                1          33.92           5                  N   
6064                 1          32.72           3                  N   
10291                1          31.95           4                  N   
29                   1          30.83           1                  N   
18130                1          30.50           1                  N   
5792                 1          30.33           2                  N   
15350                1          28.23           2                  N   
10302                1          28.20           2                  N   
2592                 1          27.97           2                  N   

       PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
9280            132           265             2       150.00    0.0      0.0   
13861           229           265             1       200.01    0.0      0.5   
6064            138             1             1       107.00    0.0      0.0   
10291           138           265             2       131.00    0.0      0.5   
29              132            23             1        80.00    0.5      0.5   
18130           132           220             1        90.50    0.0      0.5   
5792            132           158             1        52.00    0.0      0.5   
15350            13           132             1        52.00    0.0      0.5   
10302            90           132             1        52.00    0.0      0.5   
2592            261           132             2        52.00    4.5      0.5   

       tip_amount  tolls_amount  improvement_surcharge  total_amount  
9280         0.00          0.00                    0.3        150.30  
13861       51.64          5.76                    0.3        258.21  
6064        55.50         16.26                    0.3        179.06  
10291        0.00          0.00                    0.3        131.80  
29          18.56         11.52                    0.3        111.38  
18130       19.85          8.16                    0.3        119.31  
5792        14.64          5.76                    0.3         73.20  
15350        4.40          5.76                    0.3         62.96  
10302       11.71          5.76                    0.3         70.27  
2592         0.00          5.76                    0.3         63.06

In [15]:
# sort the data by total amount

total_amount_sorted = df.sort_values(
    ['total_amount'], ascending = False)['total_amount']
total_amount_sorted.head(10)

8476     1200.29
20312     450.30
13861     258.21
12511     233.74
15474     211.80
6064      179.06
16379     157.06
3582      152.30
11269     151.82
9280      150.30
Name: total_amount, dtype: float64

In [16]:
# total amount bottom 10 values
total_amount_sorted.tail(10)

5448      -4.3
10281     -4.3
4423      -4.3
20317     -4.8
8204      -4.8
11204     -5.3
14714     -5.3
17602     -5.8
20698     -5.8
12944   -120.3
Name: total_amount, dtype: float64

In [17]:
# identify payment types that are represented in the data
df['payment_type'].value_counts()

payment_type
1    15265
2     7267
3      121
4       46
Name: count, dtype: int64

From the data dictionary that the course provide, the payment method are as follows:

1 = Credit card\
2 = Cash\
3 = No Charge\
4 = Dispute\
5 = Unknown\
6 = Voided trip

In [19]:
# identify the average tip for trips paid with credit card
avg_cc_tip = df[df['payment_type'] == 1]['tip_amount'].mean()
print('Avg. cc tip:', avg_cc_tip)

# identify the average tip for trips paid with cash
avg_cash_tip = df[df['payment_type'] == 2]['tip_amount'].mean()
print('Avg. cash tip:', avg_cash_tip)

Avg. cc tip: 2.7298001965280054
Avg. cash tip: 0.0


In [20]:
# How many times is each Vendor ID represented in the data?
df['VendorID'].value_counts()

VendorID
2    12626
1    10073
Name: count, dtype: int64

In [21]:
# What is the mean total amount for each vendor?
df.groupby(['VendorID']).mean(numeric_only = True)[['total_amount']]

total_amount
VendorID              
1            16.298119
2            16.320382

In [23]:
# filter the data for credit card payments only
credit_card = df[df['payment_type'] == 1]

# filter the credit-card-only data for passenger count only
credit_card['passenger_count'].value_counts()

passenger_count
1    10977
2     2168
5      775
3      600
6      451
4      267
0       27
Name: count, dtype: int64

In [24]:
# calculate the average tip amount for each passenger count (credit card payments only)
credit_card.groupby(['passenger_count']).mean(numeric_only = True)[['tip_amount']]

tip_amount
passenger_count            
0                  2.610370
1                  2.714681
2                  2.829949
3                  2.726800
4                  2.607753
5                  2.762645
6                  2.643326

**Observation and summary**

The values for trip distance doesn't seem normal since the maximum value is 33.96 miles in about 40 minutes when looking at the datetime. Also, when sorting by total amount there seem to be an unsual value or an error as the values seem too high for a one minute trip when looking at the datetime columns. This need to bring up to the manager as the data could skew the results. Another obervation are the highest total amount doesn't relate to the highest trip distance. 

After looking at the dataset, the two varianles that are most likely to help build a predictive model for taxi ride fares are `total_amount` and `trip_distance` because those variables show a pictures of a taxi cab ride. However, the problem that was mention earlier need to be address inorder to move forward with analysis.